src/PrefixTree.java

```diff
@@ -13,7 +14,47 @@ public class PrefixTree {
     private static class TraverseResult {
         private Node node;
         private int index;
+    }
+
+    private static class Node {
+
+        private boolean endOfWord;
+        private final char value;
+        private ArrayList<Node> edges = new ArrayList<>();
+
+        Node(char value) {
+            this.value = value;
+        }
+
+        public boolean isEndOfWord() {
+            return endOfWord;
+        }
+
+        public void setEndOfWord(boolean endOfWord) {
+            this.endOfWord = endOfWord;
+        }
+
+        public char getValue() {
+            return value;
+        }
+
+        public Optional<Node> findChild(char searchedChar) {
+            return this.edges.stream()
+                    .filter(edge -> edge.getValue() == searchedChar)
+                    .findFirst();
+        }
+
+        public void addEdge(Node node) {
+            edges.add(node);
+        }
```

Comment:
Future improvement:
I wonder about this one. You can make a node point to itself and all other kind of crazy edge cases that we'll have to deal with. Won't it be better if it accepts a character and returns a new Node that's a child of this one? So `Node createChild(char value)` like. Perhaps it can even return a node if the child already exists thus removing the need to use `traverse` for adding which will get us rid of the TraverseResult. It's a big change, so it deserves its own commit after we have proper validation and some performance tests enabled.